src/cascadia/QueryExtension/GithubCopilotLLMProvider.cpp

```diff
@@ -0,0 +1,311 @@
+
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::UI::Core;
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Xaml::Controls;
+using namespace winrt::Windows::System;
+namespace WWH = ::winrt::Windows::Web::Http;
+namespace WSS = ::winrt::Windows::Storage::Streams;
+namespace WDJ = ::winrt::Windows::Data::Json;
+
+static constexpr std::wstring_view headerIconPath{ L"ms-appx:///ProfileIcons/githubCopilotLogo.png" };
+static constexpr std::wstring_view headerText{ L"GitHub Copilot" };
+static constexpr std::wstring_view subheaderText{ L"Take command of your Terminal. Ask Copilot for assistance right in your terminal." };
+static constexpr std::wstring_view badgeIconPath{ L"ms-appx:///ProfileIcons/githubCopilotBadge.png" };
+static constexpr std::wstring_view responseMetaData{ L"GitHub Copilot" };
+
+namespace winrt::Microsoft::Terminal::Query::Extension::implementation
+{
+    GithubCopilotBranding::GithubCopilotBranding()
+    {
+        _HeaderIconPath = headerIconPath;
+        _HeaderText = headerText;
+        _SubheaderText = subheaderText;
+        _BadgeIconPath = badgeIconPath;
+        _ResponseMetaData = responseMetaData;
+    }
+
+    void GithubCopilotLLMProvider::SetAuthentication(const Windows::Foundation::Collections::ValueSet& authValues)
+    {
+        _httpClient = winrt::Windows::Web::Http::HttpClient{};
+        _httpClient.DefaultRequestHeaders().Accept().TryParseAdd(L"application/json");
+        _httpClient.DefaultRequestHeaders().Append(L"Copilot-Integration-Id", L"windows-terminal-chat");
+
+        const auto url = unbox_value_or<hstring>(authValues.TryLookup(L"url").try_as<IPropertyValue>(), L"");
+        _authToken = unbox_value_or<hstring>(authValues.TryLookup(L"access_token").try_as<IPropertyValue>(), L"");
+        _refreshToken = unbox_value_or<hstring>(authValues.TryLookup(L"refresh_token").try_as<IPropertyValue>(), L"");
+
+        if (!url.empty())
+        {
+            const Windows::Foundation::Uri parsedUrl{ url };
+            const auto randomStateString = unbox_value_or<hstring>(authValues.TryLookup(L"state").try_as<IPropertyValue>(), L"");
+            // only handle this if the state strings match
+            if (randomStateString == parsedUrl.QueryParsed().GetFirstValueByName(L"state"))
+            {
+                // we got a valid URL, fire off the URL auth flow
+                _completeAuthWithUrl(parsedUrl);
+            }
+        }
+        else if (!_authToken.empty() && !_refreshToken.empty())
+        {
+            // we got tokens, use them
+            _httpClient.DefaultRequestHeaders().Authorization(WWH::Headers::HttpCredentialsHeaderValue{ L"Bearer", _authToken });
+            _obtainUsernameAndRefreshTokensIfNeeded();
+        }
+    }
+
+    winrt::fire_and_forget GithubCopilotLLMProvider::_obtainUsernameAndRefreshTokensIfNeeded()
+    {
```

Comment:
Don't forget to hold onto `this`. Consider using `safe_void_coroutine` for another layer of safety against spurious exceptions.